Improve tuning guide (#3155)
uyuni-project/uyuni-docs: modules/specialized-guides/pages/salt/salt-connectivity.adoc

[[salt-connectivity]]
= Salt Connectivity

Depending on the size and spreadness of the environment where {productname} is used some connectivity issues are possible.
There are no common recommendations for all the possible use cases as the environments could be very different expecially if public clouds instances are involved.

[[minions-connectivity]]
== Minions Connectivity

> In case if clients are losing the connection to the {productname} Server (or {productname} Proxy if involved),
> they are unreachable from the {productname} Server web UI or using the CLI.

Reviewer: ```suggestion
In case clients are losing the connection to the {productname} Server (or {productname} Proxy if involved), they are unreachable from the {productname} Server {webui} or using the CLI.
```

> To ensure if it's the case, it's better to check if the client is not reachable from the {productname} Server with [command]``salt MINION_ID test.ping``,
> while [command]``venv-salt-call test.ping`` or ([command]``salt-call test.ping`` in case if non-bundle salt is used on the client)
> is working fine on the client side, it is recommended to set `tcp_keepalive` parameters.

Reviewer: Sentence is too long.  I'm not sure if I got it right.  Maybe, we can try sth like this:

```suggestion
In such a case, first check the connection from the {productname} Server with [command]``salt MINION_ID test.ping``.
[command]``venv-salt-call test.ping`` or ([command]``salt-call test.ping`` in case non-bundle {salt} is used on the client) is working fine on the client side, especially if `tcp_keepalive` parameters are set.
```

Author: Looks like the meaning has changed here. The idea here that the minion service can lose the connection while didn't get any notification about it, so minion is still holds dead tcp socket which can't be used and it's not trying to reconnect. To ensure if it's the case one command from the server should fail while the commands `...salt-call` from the client is not failing, and in this case it makes sense to add `tcp_keepalive...` parameters to the minion config to avoid such cases.

Reviewer: I think I understand now want it is about.  Maybe, I'll come up with a proposal later.

Reviewer: Next try:
```suggestion
To debug such a situation, check whether both communication directions between server and client are working.
On the server, check with [command]``salt MINION_ID test.ping`` that the server can reach the client.
If it fails, check on the client with [command]``venv-salt-call test.ping`` whether the server is reachable from the client.
If the communication from the client is working fine, it is recommended to set `tcp_keepalive` parameters on the client.

[NOTE]
====
In case the salt bundle is not used on the client, run [command]``salt-call test.ping`` instead of [command]``venv-salt-call test.ping``.
====
```
Maybe, we should also add the technical details about the dead TCP socket?  Maybe, instead of "communication" better the simpler term "connection"?

Reviewer: Yes, is the most important thing here.  With the following proposal I try to keep the test scenario, and mostly only change the intro phrase.  Then I propose the solution in a short final sentence.  We could add the background info you provide, starting from "2 parallel channels with the same direction open. And the idea of the check is to ... etc.".

Or we just merge it as-is now, and then ask Joseph for help.

> The parameters with the values from the example below could be used as a starting point to look for the better combination, which could work better for certain environment.

Reviewer: ```suggestion
The parameters with the values from the example below could be used as a starting point to look for the better combination, which could improve performance for some environments.
```

Author: It's not about the performance, but mostly about improving the connection to prevent the cases of connection loss with no restoring.

Reviewer: Ok, I see.  Maybe, "could stabilize connection"  or "could stabilize server-client connection" or "could make connection stable".  Some with "avoid interruption"?  If you think all this is misleading, we can keep your initial wording "work better".  (I wanted to get rid of the two "betters" in one long sentence.)

Reviewer: I now think it is better to use your words.

> It's recommended to put the parameters in the separate drop-in config file like [path]``/etc/venv-salt-minion/minion.d/tuning-keepalives.conf`` or [path]``/etc/salt/minion.d/tuning-keepalives.conf``, depending on the minion type used on the client side.

Reviewer: ```suggestion
It's recommended to put the parameters in the separate drop-in configuration file like [path]``/etc/venv-salt-minion/minion.d/tuning-keepalives.conf`` or [path]``/etc/salt/minion.d/tuning-keepalives.conf``, depending on the minion type used on the client side.
```

# a state tracking device such as a firewall or VPN gateway, there is
# the risk that it could tear down the connection the master and minion
# without informing either party that their connection has been taken away.
# Enabling TCP Keepalives prevents this from happening.

# Overall state of TCP Keepalives, enable (1 or True), disable (0 or False)
# or leave to the OS defaults (-1), on Linux, typically disabled. Default True, enabled.
tcp_keepalive: True

# How long before the first keepalive should be sent in seconds. Default 300
# to send the first keepalive after 5 minutes, OS default (-1) is typically 7200 seconds
# on Linux see /proc/sys/net/ipv4/tcp_keepalive_time.
tcp_keepalive_idle: 10

# How many lost probes are needed to consider the connection lost. Default -1
# to use OS defaults, typically 9 on Linux, see /proc/sys/net/ipv4/tcp_keepalive_probes.
tcp_keepalive_cnt: 3

# How often, in seconds, to send keepalives after the first one. Default -1 to
# use OS defaults, typically 75 seconds on Linux, see
# /proc/sys/net/ipv4/tcp_keepalive_intvl.
tcp_keepalive_intvl: 10
----

[[proxies-connectivity]]
== Proxies Connectivity

> [command]``salt-broker`` service is used on {productname} Proxies to forward `salt` traffic between the {productname} Server
> and [command]``salt-minion`` services used on the client side.
> It's possible that [command]``salt-broker`` and all the clients behind it could be affected with the same issue as the clients
> directly connected to the {productname} Server.
> The issue could be fixed with the same parameters as recommended for the minions, but specified in [path]``/etc/salt/broker`` on each {productname} Proxy.

Reviewer: ```suggestion
[command]``salt-broker`` service is used on {productname} Proxies to forward `salt` traffic between the {productname} Server and [command]``salt-minion`` services used on the client side.
It is possible that [command]``salt-broker`` and all the clients behind it could be affected with the same issue as the clients directly connected to the {productname} Server.
The issue could be fixed with the same parameters as recommended for the minions, but specified in [path]``/etc/salt/broker`` on each {productname} Proxy.
```

Author: ```suggestion
[command]``salt-broker`` service is used on {productname} Proxies to forward `salt` traffic between the {productname} Server and [command]``salt-minion`` service used on the client side.
It is possible that [command]``salt-broker`` and all the clients behind it could be affected with the same issue as the clients directly connected to the {productname} Server.
The issue could be fixed with the same parameters as recommended for the minions, but specified in [path]``/etc/salt/broker`` on each {productname} Proxy.
```

Author: Just 1 small change from your suggestion, just noticed not really good typo

The other possible issue which {productname} Proxy can be affected with is the case when the connectivity to the {productname} Server
is lost for quite long interval, so the clients behind it started to retry the authentication to the [command]``salt-master`` service on the {productname} Server.
This situation could be potentially dangerous as it could lead to collecting large amount of ZeroMQ messages with authentication attemps
in the interal buffer of ZeroMQ sockets used inside the [command]``salt-broker`` service, so that on restoring the connection to the [command]``salt-master``

> all of the messages will be pushed to it. It could lead to the issues on {productname} Server side with [command]``salt-master`` service,
> as it could be impossible to serve all the cached requests in apropriate time or even to the complete denial of the service.

Reviewer: ```suggestion
all of the messages will be pushed to it.
It could lead to the issues on {productname} Server side with [command]``salt-master`` service, as it could be impossible to serve all the cached requests in apropriate time or even to the complete denial of the service.
```

> To avoid such situation a set of extra parameters was introduced, the most important one is ``wait_for_backend``, which sould be set to ``True``,
> it prevents opening the sockets for the clients behind the proxy while the connectivity to the [command]``salt-master`` service is not established.

Reviewer: ```suggestion
To avoid such situation a set of extra parameters was introduced.
The most important one is [option]``wait_for_backend``, which should be set to [literal]``True``.
This prevents opening the sockets for the clients behind the proxy while the connectivity to the [command]``salt-master`` service is not established.
```

In this case the messages from the clients are not collected in the internal buffers.
``drop_after_retries`` is setting the number of retries before closing the sockets to drop the cached messages.
The other parameters could help to fine tune the behaviour for the environment.

[NOTE]
====

> Setting timouts, intevals and ``drop_after_retries`` to lower values are making [command]``salt-broker`` service more agressive on detecting
> the connection loss to the [command]``salt-master``, so that it puts the clients behind the proxy to the conditions closer as they are connected

Reviewer: ```suggestion
Setting timeouts, intervals and ``drop_after_retries`` to lower values are making [command]``salt-broker`` service more aggressive on detecting the connection loss to the [command]``salt-master``, so that it puts the clients behind the proxy to the conditions closer as they are connected
```
